File: ffupdater/src/main/java/de/marmaro/krt/ffupdater/metadata/fetcher/GithubConsumer.java
package de.marmaro.krt.ffupdater.metadata.fetcher;

import androidx.annotation.NonNull;

import com.google.common.base.Preconditions;
import com.google.gson.annotations.SerializedName;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.function.Predicate;

import de.marmaro.krt.ffupdater.utils.ParamRuntimeException;
import de.marmaro.krt.ffupdater.utils.Utils;

public class GithubConsumer {
    public static final String LATEST_RELEASE_URL = "https://api.github.com/repos/%s/%s/releases/latest";
    public static final String ALL_RELEASES_URL = "https://api.github.com/repos/%s/%s/releases?per_page=%d&page=%d";

    private final ApiConsumer apiConsumer;

    public GithubConsumer(ApiConsumer apiConsumer) {
        Objects.requireNonNull(apiConsumer);
        this.apiConsumer = apiConsumer;
    }

    GithubResult consumeLatestReleaseFirst(Request request) {
        Preconditions.checkArgument(isRequestValid(request));
        final URL url = Utils.createURL(String.format(
                LATEST_RELEASE_URL,
                request.ownerOfRepository,
                request.repositoryName));
        Release release = apiConsumer.consume(url, Release.class);
        if (request.releaseValidator.test(release)) {
            return convert(release);
        }
        return consumeManyReleases(request);
    }

    GithubResult consumeManyReleases(Request request) {
        Preconditions.checkArgument(isRequestValid(request));
        for (int page = 1; page < 5; page++) {
            final URL url = Utils.createURL(String.format(
                    Locale.getDefault(),
                    ALL_RELEASES_URL,
                    request.ownerOfRepository,
                    request.repositoryName,
                    request.resultsPerPage,
                    page));
            final Release[] test = apiConsumer.consume(url, Release[].class);
            for (Release release : test) {
                if (request.releaseValidator.test(release)) {
                    return convert(release);
                }
            }
        }
        throw new ParamRuntimeException("can't find release after nine network requests - abort");
    }

    private boolean isRequestValid(Request request) {
        return request != null
                && request.ownerOfRepository != null
                && request.repositoryName != null
                && request.resultsPerPage != -1
                && request.releaseValidator != null;
    }

    private GithubResult convert(Release release) {
        final Map<String, URL> urls = new HashMap<>();
        try {
            for (Asset asset : release.getAssets()) {
                urls.put(asset.getName(), new URL(asset.getDownloadUrl()));
            }
        } catch (MalformedURLException e) {
            throw new ParamRuntimeException("invalid urls");
        }
        return new GithubResult(release.getTagName(), urls);
    }

    public static class Request {
        private String ownerOfRepository;
        private String repositoryName;
        private int resultsPerPage = -1;
        private Predicate<Release> releaseValidator;

        public Request setOwnerOfRepository(String ownerOfRepository) {
            this.ownerOfRepository = ownerOfRepository;
            return this;
        }

        public Request setRepositoryName(String repositoryName) {
            this.repositoryName = repositoryName;
            return this;
        }

        public Request setResultsPerPage(int resultsPerPage) {
            this.resultsPerPage = resultsPerPage;
            return this;
        }

        public Request setReleaseValidator(Predicate<Release> releaseValidator) {
            this.releaseValidator = releaseValidator;
            return this;
        }
    }

    static class GithubResult {
        private final String tagName;
        private final Map<String, URL> urls;

        GithubResult(String tagName, Map<String, URL> urls) {
            this.tagName = tagName;
            this.urls = urls;
        }

        String getTagName() {
            return tagName;
        }

        Map<String, URL> getUrls() {
            return urls;
        }
    }

    static class Release {
        @SerializedName("tag_name")
        private String tagName;

        @SerializedName("name")
        private String name;

        @SerializedName("prerelease")
        private boolean prerelease;

        @SerializedName("assets")
        private List<Asset> assets;

        public String getTagName() {
            return tagName;
        }

        public String getName() {
            return name;
        }

        public boolean isPrerelease() {
            return prerelease;
        }

        public List<Asset> getAssets() {
            return assets;
        }

        @NonNull
        @Override
        public String toString() {
            return "Release{" +
                    "tagName='" + tagName + '\'' +
                    ", assets=" + assets +
                    '}';
        }
    }

    static class Asset {
        @SerializedName("name")
        private String name;

        @SerializedName("browser_download_url")
        private String downloadUrl;

        String getName() {
            return name;
        }

        String getDownloadUrl() {
            return downloadUrl;
        }

        @NonNull
        @Override
        public String toString() {
            return "Asset{" +
                    "name='" + name + '\'' +
                    ", downloadUrl='" + downloadUrl + '\'' +
                    '}';
        }
    }
}

File: ffupdater/src/main/java/de/marmaro/krt/ffupdater/app/impl/fetch/github/GithubConsumer.kt
package de.marmaro.krt.ffupdater.app.impl.fetch.github

import androidx.annotation.MainThread
import com.google.gson.annotations.SerializedName
import de.marmaro.krt.ffupdater.app.impl.exceptions.InvalidApiResponseException
import de.marmaro.krt.ffupdater.app.impl.fetch.ApiConsumer
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter.ISO_ZONED_DATE_TIME
import java.util.*
import java.util.function.Predicate

class GithubConsumer(
    private val repoOwner: String,
    private val repoName: String,
    private val resultsPerPage: Int,
    private val isValidRelease: Predicate<Release>,
    private val isCorrectAsset: Predicate<Asset>,
) {
    init {
        check(resultsPerPage > 0)
    }

    @MainThread
    suspend fun updateCheck(): Result {
        val url = "https://api.github.com/repos/$repoOwner/$repoName/releases/latest"
        val release = ApiConsumer.consumeNetworkResource(url, Release::class)
        if (isValidRelease.test(release)) {
            return convert(release)
        }
        return updateCheckAllReleases()
    }

    @MainThread
    suspend fun updateCheckAllReleases(): Result {
        val tries = 4
        for (page in 1..(tries + 1)) {
            val url = "https://api.github.com/repos/$repoOwner/$repoName/releases" +
                    "?per_page=$resultsPerPage&page=$page"
            val releases = ApiConsumer.consumeNetworkResource(url, Array<Release>::class)
            releases.forEach {
                if (isValidRelease.test(it)) {
                    return convert(it)
                }
            }
        }
        throw InvalidApiResponseException("can't find release after $tries tries - abort")
    }

    /**
     * @throws InvalidApiResponseException
     */
    private fun convert(release: Release): Result {
        val asset = release.assets.firstOrNull { isCorrectAsset.test(it) }
                ?: throw InvalidApiResponseException("${release.name} has no suitable asset")
        return Result(
                tagName = release.tagName,
                url = asset.downloadUrl,
                fileSizeBytes = asset.fileSizeBytes,
                releaseDate = ZonedDateTime.parse(release.publishedAt, ISO_ZONED_DATE_TIME))
    }

    data class Release(
            @SerializedName("tag_name")
            val tagName: String,
            @SerializedName("name")
            val name: String,
            @SerializedName("prerelease")
            val isPreRelease: Boolean,
            @SerializedName("assets")
            val assets: List<Asset>,
            @SerializedName("published_at")
            val publishedAt: String,
    )

    data class Asset(
            @SerializedName("name")
            val name: String,
            @SerializedName("browser_download_url")
            val downloadUrl: String,
            @SerializedName("size")
            val fileSizeBytes: Long,
    )

    data class Result(
            val tagName: String,
            val url: String,
            val fileSizeBytes: Long,
            val releaseDate: ZonedDateTime,
    )
}